File: src/lambda/dms-switch/dms-post.ts
import { CloudFormationClient } from "@aws-sdk/client-cloudformation";
import {
  DatabaseMigrationServiceClient,
  StartReplicationTaskCommand,
} from "@aws-sdk/client-database-migration-service";
import type {
  CloudFormationCustomResourceEvent,
  CloudFormationCustomResourceFailedResponse,
  CloudFormationCustomResourceSuccessResponse,
} from "aws-lambda";
import { getDmsTaskStatus } from "./utils/get-dms-status";
import { waitForDmsStatus } from "./utils/wait-for-dms-status";
import { hasDmsChanges } from "./utils/has-dms-changes";

const dms = new DatabaseMigrationServiceClient({});
const cf = new CloudFormationClient({});

export const handler = async (
  event: CloudFormationCustomResourceEvent
): Promise<
  | CloudFormationCustomResourceSuccessResponse
  | CloudFormationCustomResourceFailedResponse
> => {
  // note Delete shouldn't re-start DMS since the stack is being deleted
  const ReplicationTaskArn = `${process.env.DMS_TASK}`;
  console.log(JSON.stringify({ RequestType: event.RequestType }));
  try {
    switch (event.RequestType) {
      case "Create":
        const startCmd = new StartReplicationTaskCommand({
          ReplicationTaskArn,
          StartReplicationTaskType: "start-replication",
        });
        await dms.send(startCmd);
        await waitForDmsStatus({
          dms,
          ReplicationTaskArn,
          targetStatus: "running",
        });
        return { ...event, PhysicalResourceId: "post-dms", Status: "SUCCESS" };
      case "Update":
        let shouldUnpause = false;
        const dmsChanges = await hasDmsChanges({cf,
          StackName: `${process.env.STACK_NAME}`,
        });
        if (dmsChanges) {
          shouldUnpause = true;
        } else {
          const status = await getDmsTaskStatus({ dms, ReplicationTaskArn });
          console.log(`DMS status: ${status}`);
          if (status === "stopped" || status === "ready") {
            shouldUnpause = true;
          }
        }

        if (shouldUnpause) {
          // unpause DMS
          const startCmd = new StartReplicationTaskCommand({
            ReplicationTaskArn,
            StartReplicationTaskType: "resume-processing",
          });
          await dms.send(startCmd);
          await waitForDmsStatus({
            dms,
            ReplicationTaskArn,
            targetStatus: "running",
          });
        }
        return { ...event, PhysicalResourceId: "post-dms", Status: "SUCCESS" };

      default:
        console.error("No op for", event.RequestType);
        return { ...event, PhysicalResourceId: "post-dms", Status: "SUCCESS" };
    }
  } catch (e) {
    console.error(`Failed!`, e);
    return {
      ...event,
      PhysicalResourceId: "post-dms",
      Reason: (e as Error).message,
      Status: "FAILED",
    };
  }
};

File: src/lambda/dms-switch-py/dms_post.py
import os
import json
import boto3
from utils.wait_for_dms_status import wait_for_dms_status
from utils.has_dms_changes import has_dms_changes
from utils.get_dms_task_status import get_dms_task_status


dms = boto3.client('dms')
cf = boto3.client('cloudformation')


def handler(event, context):
    replication_task_arn = os.environ.get('DMS_TASK')
    print(json.dumps({'RequestType': event['RequestType']}))
    try:
        if event['RequestType'] == 'Create':
            start_cmd = {
                'ReplicationTaskArn': replication_task_arn,
                'StartReplicationTaskType': 'start-replication'
            }
            dms.start_replication_task(**start_cmd)
            wait_for_dms_status(replication_task_arn, 'running')
            return {
                'PhysicalResourceId': 'post-dms',
                'Status': 'SUCCESS'
            }
        elif event['RequestType'] == 'Update':
            should_unpause = False
            dms_changes = has_dms_changes(
                event['ResourceProperties']['StackName'])
            if dms_changes:
                should_unpause = True
            else:
                status = get_dms_task_status(replication_task_arn)
                print(f'DMS status: {status}')
                if status in ('stopped', 'ready'):
                    should_unpause = True
            if should_unpause:
                start_cmd = {
                    'ReplicationTaskArn': replication_task_arn,
                    'StartReplicationTaskType': 'resume-processing'
                }
                dms.start_replication_task(**start_cmd)
                wait_for_dms_status(replication_task_arn, 'running')
            return {
                'PhysicalResourceId': 'post-dms',
                'Status': 'SUCCESS'
            }
        else:
            print('No operation for', event['RequestType'])
            return {
                'PhysicalResourceId': 'post-dms',
                'Status': 'SUCCESS'
            }
    except Exception as e:
        print('Failed!', str(e))
        return {
            'PhysicalResourceId': 'post-dms',
            'Reason': str(e),
            'Status': 'FAILED'
        }
